src/serializers/type_serializers/dataclass.rs

```diff
@@ -54,16 +58,137 @@ impl BuildSerializer for DataclassArgsBuilder {
     }
 }
 
-pub struct DataclassBuilder;
+#[derive(Debug, Clone)]
+pub struct DataclassSerializer {
+    class: Py<PyType>,
+    serializer: Box<CombinedSerializer>,
+    fields: Vec<Py<PyString>>,
+    name: String,
+}
 
-impl BuildSerializer for DataclassBuilder {
+impl BuildSerializer for DataclassSerializer {
     const EXPECTED_TYPE: &'static str = "dataclass";
 
     fn build(
         schema: &PyDict,
-        config: Option<&PyDict>,
+        _config: Option<&PyDict>,
         definitions: &mut DefinitionsBuilder<CombinedSerializer>,
     ) -> PyResult<CombinedSerializer> {
-        ModelSerializer::build(schema, config, definitions)
+        let py = schema.py();
+
+        // models ignore the parent config and always use the config from this model
+        let config = schema.get_as(intern!(py, "config"))?;
+
+        let class: &PyType = schema.get_as_req(intern!(py, "cls"))?;
+        let sub_schema: &PyDict = schema.get_as_req(intern!(py, "schema"))?;
+        let serializer = Box::new(CombinedSerializer::build(sub_schema, config, definitions)?);
+
+        let dc_fields: &PyDict = class.getattr(intern!(py, "__dataclass_fields__"))?.downcast()?;
+        let mut fields = Vec::with_capacity(dc_fields.len());
+
+        let field_type_marker = get_field_marker(py)?;
+        for (field_name, field) in dc_fields.iter() {
+            let field_type = field.getattr(intern!(py, "_field_type"))?;
+            if field_type.is(field_type_marker) {
+                let field_name: &PyString = field_name.downcast()?;
+                fields.push(field_name.into_py(py));
+            }
+        }
+
+        Ok(Self {
+            class: class.into(),
+            serializer,
+            fields,
+            name: class.getattr(intern!(py, "__name__"))?.extract()?,
+        }
+        .into())
+    }
+}
+
+impl DataclassSerializer {
+    fn allow_value(&self, value: &PyAny, extra: &Extra) -> PyResult<bool> {
+        match extra.check {
+            SerCheck::Strict => Ok(value.get_type().is(self.class.as_ref(value.py()))),
+            SerCheck::Lax => value.is_instance(self.class.as_ref(value.py())),
+            SerCheck::None => value.hasattr(intern!(value.py(), "__dataclass_fields__")),
+        }
+    }
+
+    fn get_inner_value<'py>(&self, value: &'py PyAny) -> PyResult<&'py PyAny> {
+        let py = value.py();
+        let dict = PyDict::new(py);
+
+        for field_name in &self.fields {
```

Comment:
I think maybe worth a comment here explaining why we aren't just using `__dict__`

Reply:
done.